src/test/app/Check_test.cpp

```diff
@@ -1848,22 +1883,654 @@ class Check_test : public beast::unit_test::suite
+                cmpReqAmount(sfHighLimit);
+            }
+            {
+                // Lambda that compares the contents of optional fields.
+                auto cmpOptField = [this, offerLine, checkLine](
+                                       auto const& sfield) {
+                    // Expect both fields to either be present or absent.
+                    BEAST_EXPECT(
+                        offerLine->isFieldPresent(sfield) ==
+                        checkLine->isFieldPresent(sfield));
+
+                    if (!offerLine->isFieldPresent(sfield) ||
+                        !checkLine->isFieldPresent(sfield))
+                        return;
+
+                    // Both optional fields are present so we can compare them.
+                    BEAST_EXPECT((*offerLine)[sfield] == (*checkLine)[sfield]);
+                };
+                cmpOptField(sfLowNode);
+                cmpOptField(sfLowQualityIn);
+                cmpOptField(sfLowQualityOut);
+
+                cmpOptField(sfHighNode);
+                cmpOptField(sfHighQualityIn);
+                cmpOptField(sfHighQualityOut);
+            }
+        };
+
+        //----------- No account root flags, check written by issuer -----------
+        {
+            // No account root flags on any participant.
+            // Automatic trust line from issuer to destination.
+
+            // Use offers to automatically create the trust line.
+            IOU const OF1 = gw1["OF1"];
+            env(offer(gw1, XRP(98), OF1(98)));
+            env.close();
+            BEAST_EXPECT(
+                env.le(keylet::line(gw1, alice, OF1.currency)) == nullptr);
+            env(offer(alice, OF1(98), XRP(98)));
+            env.close();
+
+            // Both offers should be consumed.
+            // Since gw1's offer was consumed and the trust line was not
+            // created by gw1, gw1's owner count should be 0.
+            BEAST_EXPECT(ownerCount(env, gw1) == 0);
+
+            // alice's automatically created trust line bumps her owner count.
+            BEAST_EXPECT(ownerCount(env, alice) == 1);
+
+            // Use check cashing to automatically create the trust line.
+            IOU const CK1 = gw1["CK1"];
+            uint256 const chkId{getCheckIndex(gw1, env.seq(gw1))};
+            env(check::create(gw1, alice, CK1(98)));
+            env.close();
+            BEAST_EXPECT(
+                env.le(keylet::line(gw1, alice, CK1.currency)) == nullptr);
+            env(check::cash(alice, chkId, CK1(98)));
+            verifyDeliveredAmount(env, CK1(98));
+            env.close();
+
+            // gw1's check should be consumed.
+            // Since gw1's check was consumed and the trust line was not
+            // created by gw1, gw1's owner count should be 0.
+            BEAST_EXPECT(ownerCount(env, gw1) == 0);
+
+            // alice's automatically created trust line bumps her owner count.
+            BEAST_EXPECT(ownerCount(env, alice) == 2);
```

Comment:
Overall, these tests are _fantastic_.
One thing I think could help with readability, would be to use a set of integers to keep track of what each account's owner count _should_ be. Then increment (or decrement) that value whenever a change is expected. This would allow the reader to look at each test case separately without needing to flip up to the previous test to check the prior value. Down the road, it would also allow reordering the test cases or inserting new test cases between the existing ones.
e.g.
```
        Account const gw1{"gw1"};
        std::size_t gw1Count = 0;
        Account const alice{"alice"};
        std::size_t aliceCount = 0;
        Account const bob{"bob"};
        std::size_t bobCount = 0;
[...]
            // gw1's check should be consumed.
            // Since gw1's check was consumed and the trust line was not
            // created by gw1, gw1's owner count should be 0.
            BEAST_EXPECT(ownerCount(env, gw1) == gw1Count);

            // alice's automatically created trust line bumps her owner count.
            BEAST_EXPECT(ownerCount(env, alice) == ++aliceCount);
```